README
SWH-loader-tar
==============

The Software Heritage Tarball Loader is a tool and a library to
uncompress a local tarball and inject into the SWH dataset all unknown
contained files.

Tarball loader
==============

Its job is to uncompress a tarball and load its content in swh
storage.

### Configuration

This is the loader's (or task's) configuration file.

loader/tar.yml:

```
    extraction_dir: /home/storage/tmp/
    storage:
      cls: local
      args:
        db: service=swh-dev
        objstorage:
          cls: pathslicing
          args:
            root: /home/storage/swh-storage
            slicing: 0:2/2:4/4:6
    send_contents: True
    send_directories: True
    send_revisions: True
    send_releases: True
    send_occurrences: True

    content_packet_size: 10000
    content_packet_block_size_bytes: 104857600
    content_packet_size_bytes: 1073741824
    directory_packet_size: 25000
    revision_packet_size: 100000
    release_packet_size: 100000
    occurrence_packet_size: 100000
    extraction_dir = /home/storage/tmp/
```

Present in possible locations:
- ~/.config/swh/loader/tar.ini
- ~/.swh/loader/tar.ini
- /etc/softwareheritage/loader/tar.ini


### API

Load tarball directly from code or toplevel:

    from swh.loader.tar.loader import TarLoader

    tarpath = '/some/path/to/blah-7.8.3.tgz'
    # Fill in those
    origin = {'url': 'some-origin', 'type': 'dir'}
    visit_date = 'Tue, 3 May 2017 17:16:32 +0200'
    revision = {}
    occurrence = {}

    TarLoader().load(tarpath, origin, visit_date, revision, [occurrence])


### Celery

Load tarball using celery.

Providing you have a properly configured celery up and running

worker.ini needs to be updated with the following keys:

    task_modules = swh.loader.tar.tasks
    task_queues = swh_loader_tar

cf. https://forge.softwareheritage.org/diffusion/DCORE/browse/master/README.md
for more details


#### Toplevel

You can send the following message to the task queue:

    from swh.loader.tar.tasks import LoadTarRepository

    # Fill in those
    tarpath = '/some/path/to/blah-7.8.3.tgz'
    origin = {'url': 'some-origin', 'type': 'dir'}
    visit_date = 'Tue, 3 May 2017 17:16:32 +0200'
    revision = {}
    occurrence = {}

    # Send message to the task queue
    LoadTarRepository().run((tarpath, origin, visit_date, revision, [occurrence]))


Tar Producer
============

Its job is to compulse from a file or a folder a list of existing
tarballs. From this list, compute the corresponding messages to send
to the broker.

#### Configuration

Message producer's configuration file:

    [main]

    # Mirror's root directory holding tarballs to load into swh
    mirror_root_directory = /srv/storage/space/mirrors/gnu.org/gnu/
    # mirror_root_directory = /srv/storage/space/mirrors/gnu.org/old-gnu/

    # Url scheme prefix used to create the origin url
    url_scheme = http://ftp.gnu.org/gnu/
    # url_scheme = rsync://ftp.gnu.org/old-gnu/

    # Origin type used for tarballs
    type = ftp

    # File containing a subset list tarballs from mirror_root_directory to load.
    # The file's format is one absolute path name to a tarball per line.
    # NOTE:
    # - This file must contain data consistent with the mirror_root_directory
    # - if this option is not provided, the mirror_root_directory is scanned
    # completely as usual
    # mirror_subset_archives = /home/storage/missing-archives

    # Randomize blocks of messages and send for consumption
    block_messages = 250

#### Run

Trigger the message computations:

    python3 -m swh.loader.tar.producer --config ~/.swh/producer/tar.ini

This will walk the `mirror_root_directory` folder and send encountered
tarball messages for the swh-loader-tar to uncompress (through
celery).

If the `mirror_subset_archives` is provided, the tarball messages will
be computed from such file (the mirror_root_directory is still used so
please be consistent).

If problem arises during tarball message computation, a message will be
outputed with the tarball that present a problem.

It will displayed the number of tarball messages sent at the end.

Dry run:

    python3 -m swh.loader.tar.producer --config-file ~/.swh/producer/tar.ini --dry-run

This will do the same as previously described but only display the
number of potential tarball messages computed.

Help:

    python3 -m swh.loader.tar.producer --help


diff-db-mirror
==============

Utility to compute the difference between the `occurrence_history` table
(column branch) and the actual mirror path on disk.
This will output the path to the tarballs not injected in db (for any reason).

This output is to be consumed by the swh-loader-tar-producer in replay mode.


Sample use:

    ./bin/swh-diff-db-mirror \
      --db-url 'host=<host> dbname=<db> user=<user> password=<pass>' \
      --mirror-root-directory /path/to/mirrors/gnu.org/old-gnu

Here is a sample output:

    ...
    /home/storage/space/mirrors/gnu.org/gnu/miscfiles/miscfiles-1.4.2.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.4.5.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.10.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.4.8.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.5.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.7.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.14.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.2.59.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.9.tar.gz
    /home/storage/space/mirrors/gnu.org/gnu/zile/zile-2.3.11.tar.gz


debian/changelog
swh-loader-tar (0.0.41-1~swh1~bpo10+1) buster-swh; urgency=medium

  * Rebuild for buster-swh

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Thu, 10 Oct 2019 09:49:46 +0000

swh-loader-tar (0.0.41-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.0.41     - (tagged by Stefano Zacchiroli
    <zack@upsilon.cc> on 2019-10-10 11:45:19 +0200)
  * Upstream changes:     - v0.0.41     - * tox.ini: Fix py3 environment
    to use packaged tests     - * typing: minimal changes to make a no-
    op mypy run pass     - * Add attribute visit_type.     - * add code
    of conduct document     - * CONTRIBUTORS: Add Thibault Allançon     -
    * tox: workaround to pip's inability to properly solve dependency
    resolution

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Thu, 10 Oct 2019 09:48:23 +0000

swh-loader-tar (0.0.40-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.0.40     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2019-04-19 08:51:58 +0200)
  * Upstream changes:     - Release swh.loader.tar v0.0.40     - Pass
    through metadata from the caller

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 19 Apr 2019 06:58:41 +0000

swh-loader-tar (0.0.39-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.0.39     - (tagged by Antoine R. Dumont
    (@ardumont) <antoine.romain.dumont@gmail.com> on 2019-02-18 11:44:02
    +0100)
  * Upstream changes:     - v0.0.39     - tar.loader: Fix default value
    initialization caveat

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 18 Feb 2019 10:49:32 +0000

swh-loader-tar (0.0.38-1~swh2) unstable-swh; urgency=medium

  * New upstream release, fix debian tree

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Thu, 14 Feb 2019 17:56:44 +0100

swh-loader-tar (0.0.38-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.0.38     - (tagged by David Douard
    <david.douard@sdfa3.org> on 2019-02-07 17:34:18 +0100)
  * Upstream changes:     - v0.0.38

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Thu, 07 Feb 2019 16:39:05 +0000

swh-loader-tar (0.0.35-1~swh1) unstable-swh; urgency=medium

  * v0.0.35
  * Improve origin_visit initialization step
  * Properly sandbox the prepare statement so that if it breaks, we can
  * update appropriately the visit with the correct status

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Wed, 07 Mar 2018 12:00:13 +0100

swh-loader-tar (0.0.34-1~swh1) unstable-swh; urgency=medium

  * v0.0.34
  * loader: replace occurrences with snapshot

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Mon, 12 Feb 2018 11:17:17 +0100

swh-loader-tar (0.0.33-1~swh1) unstable-swh; urgency=medium

  * Release swh.loader.tar v0.0.33
  * Use snapshots instead of occurrences

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Tue, 06 Feb 2018 14:34:47 +0100

swh-loader-tar (0.0.32-1~swh1) unstable-swh; urgency=medium

  * v0.0.32
  * Migrate to latest swh-core

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Wed, 06 Dec 2017 12:04:02 +0100

swh-loader-tar (0.0.31-1~swh1) unstable-swh; urgency=medium

  * v0.0.31
  * Bump to latest swh.loader.dir
  * Fix test

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Mon, 04 Dec 2017 18:55:50 +0100

swh-loader-tar (0.0.30-1~swh1) unstable-swh; urgency=medium

  * v0.0.30
  * Do not override metadata if already provided

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Mon, 04 Dec 2017 18:13:05 +0100

swh-loader-tar (0.0.29-1~swh1) unstable-swh; urgency=medium

  * v0.0.29
  * Permit inheritance and configuration from loader-tar

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Wed, 29 Nov 2017 14:39:35 +0100

swh-loader-tar (0.0.28-1~swh1) unstable-swh; urgency=medium

  * Release swh.loader.tar version 0.0.28
  * Update packaging runes

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Thu, 12 Oct 2017 18:07:58 +0200

swh-loader-tar (0.0.27-1~swh1) unstable-swh; urgency=medium

  * Release swh.loader.tar v0.0.27
  * Update to new swh.loader.core

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Fri, 06 Oct 2017 15:11:40 +0200

swh-loader-tar (0.0.26-1~swh1) unstable-swh; urgency=medium

  * v0.0.26
  * Bump dependency on swh.loader.dir
  * Add a one no-persistence test

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Fri, 29 Sep 2017 14:26:10 +0200

swh-loader-tar (0.0.25-1~swh1) unstable-swh; urgency=medium

  * Release swh.loader.tar 0.0.25
  * Update tasks to new swh.scheduler API

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Mon, 12 Jun 2017 18:05:35 +0200

swh-loader-tar (0.0.24-1~swh1) unstable-swh; urgency=medium

  * Release swh.loader.tar v0.0.24
  * migrate to swh.model.hashutil

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Thu, 04 May 2017 16:29:50 +0200

swh-loader-tar (0.0.23-1~swh1) unstable-swh; urgency=medium

  * v0.0.23
  * Simplify loader's logic

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Wed, 22 Feb 2017 13:50:05 +0100

swh-loader-tar (0.0.22-1~swh1) unstable-swh; urgency=medium

  * v0.0.22
  * Update storage configuration reading

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Thu, 15 Dec 2016 18:53:35 +0100

swh-loader-tar (0.0.21-1~swh1) unstable-swh; urgency=medium

  * v0.0.21
  * Update tar loader to register origin_visit's state

 -- Antoine R. Dumont (@ardumont) <antoine.romain.dumont@gmail.com>  Wed, 24 Aug 2016 14:52:57 +0200

swh-loader-tar (0.0.20-1~swh1) unstable-swh; urgency=medium

  * v0.0.20
  * Stabilize dependency on swh-loader-dir


swh/loader/tar/loader.py
# Copyright (C) 2015-2019  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU General Public License version 3, or any later version
# See top-level LICENSE file for more information


import os
import tempfile
import requests
import shutil
from urllib.parse import urlparse

from tempfile import mkdtemp

from swh.core import tarball
from swh.loader.core.loader import BufferedLoader
from swh.loader.dir.loader import revision_from, snapshot_from
from swh.model.hashutil import MultiHash, HASH_BLOCK_SIZE
from swh.model.from_disk import Directory

from .build import compute_revision, set_original_artifact

try:
    from _version import __version__
except ImportError:
    __version__ = 'devel'


TEMPORARY_DIR_PREFIX_PATTERN = 'swh.loader.tar.'
DEBUG_MODE = '** DEBUG MODE **'


class LocalResponse:
    """Local Response class with iter_content api

    """
    def __init__(self, path):
        self.path = path

    def iter_content(self, chunk_size=None):
        with open(self.path, 'rb') as f:
            for chunk in f:
                yield chunk


class ArchiveFetcher:
    """Http/Local client in charge of downloading archives from a
       remote/local server.

    Args:
        temp_directory (str): Path to the temporary disk location used
                              for downloading the release artifacts

    """
    def __init__(self, temp_directory=None):
        self.temp_directory = temp_directory
        self.session = requests.session()
        self.params = {
            'headers': {
                'User-Agent': 'Software Heritage Tar Loader (%s)' % (
                    __version__
                )
            }
        }

    def download(self, url):
        """Download the remote tarball url locally.

        Args:
            url (str): Url (file or http*)

        Raises:
            ValueError in case of failing to query

        Returns:
            Tuple of local (filepath, hashes of filepath)

        """
        url_parsed = urlparse(url)
        if url_parsed.scheme == 'file':
            path = url_parsed.path
            response = LocalResponse(path)
            length = os.path.getsize(path)
        else:
            response = self.session.get(url, **self.params, stream=True)
            if response.status_code != 200:
                raise ValueError("Fail to query '%s'. Reason: %s" % (
                    url, response.status_code))
            length = int(response.headers['content-length'])

        filepath = os.path.join(self.temp_directory, os.path.basename(url))

        h = MultiHash(length=length)
        with open(filepath, 'wb') as f:
            for chunk in response.iter_content(chunk_size=HASH_BLOCK_SIZE):
                h.update(chunk)
                f.write(chunk)

        actual_length = os.path.getsize(filepath)
        if length != actual_length:
            raise ValueError('Error when checking size: %s != %s' % (
                length, actual_length))

        hashes = {
            'length': length,
            **h.hexdigest()
        }
        return filepath, hashes


class BaseTarLoader(BufferedLoader):
    """Base Tarball Loader class.

    This factorizes multiple loader implementations:

      - :class:`RemoteTarLoader`: New implementation able to deal with
         remote archives.

      - :class:`TarLoader`: Old implementation which dealt with only
         local archive. It also was only passing along objects to
         persist (revision, etc...)

    """
    CONFIG_BASE_FILENAME = 'loader/tar'

    ADDITIONAL_CONFIG = {
        'working_dir': ('string', '/tmp'),
        'debug': ('bool', False),  # NOT FOR PRODUCTION
    }

    visit_type = 'tar'

    def __init__(self, logging_class='swh.loader.tar.TarLoader', config=None):
        super().__init__(logging_class=logging_class, config=config)
        self.local_cache = None
        self.dir_path = None
        working_dir = self.config.get('working_dir', tempfile.gettempdir())
        os.makedirs(working_dir, exist_ok=True)
        self.temp_directory = mkdtemp(
            suffix='-%s' % os.getpid(),
            prefix=TEMPORARY_DIR_PREFIX_PATTERN,
            dir=working_dir)
        self.client = ArchiveFetcher(temp_directory=self.temp_directory)
        os.makedirs(working_dir, 0o755, exist_ok=True)
        self.dir_path = tempfile.mkdtemp(prefix='swh.loader.tar-',
                                         dir=self.temp_directory)
        self.debug = self.config.get('debug', False)

    def cleanup(self):
        """Clean up temporary disk folders used.

        """
        if self.debug:
            self.log.warn('%s Will not clean up temp dir %s' % (
                DEBUG_MODE, self.temp_directory
            ))
            return
        if os.path.exists(self.temp_directory):
            self.log.debug('Clean up %s' % self.temp_directory)
            shutil.rmtree(self.temp_directory)

    def prepare_origin_visit(self, *, origin, visit_date=None, **kwargs):
        """Prepare the origin visit information.

        Args:
            origin (dict): Dict with keys {url, type}
            visit_date (str): Date representing the date of the
              visit. None by default will make it the current time
              during the loading process.

        """
        self.origin = origin
        if 'type' not in self.origin:  # let the type flow if present
            self.origin['type'] = self.visit_type
        self.visit_date = visit_date

    def get_tarball_url_to_retrieve(self):
        """Compute the tarball url to allow retrieval

        """
        raise NotImplementedError()

    def fetch_data(self):
        """Retrieve, uncompress archive and fetch objects from the tarball.
           The actual ingestion takes place in the :meth:`store_data`
           implementation below.

        """
        url = self.get_tarball_url_to_retrieve()
        filepath, hashes = self.client.download(url)
        nature = tarball.uncompress(filepath, self.dir_path)

        dir_path = self.dir_path.encode('utf-8')
        directory = Directory.from_disk(path=dir_path, save_path=True)
        objects = directory.collect()
        if 'content' not in objects:
            objects['content'] = {}
        if 'directory' not in objects:
            objects['directory'] = {}

        # compute the full revision (with ids)
        revision = self.build_revision(filepath, nature, hashes)
        revision = revision_from(directory.hash, revision)
        objects['revision'] = {
            revision['id']: revision,
        }

        snapshot = self.build_snapshot(revision)
        objects['snapshot'] = {
            snapshot['id']: snapshot
        }
        self.objects = objects

    def store_data(self):
        """Store the objects in the swh archive.

        """
        objects = self.objects
        self.maybe_load_contents(objects['content'].values())
        self.maybe_load_directories(objects['directory'].values())
        self.maybe_load_revisions(objects['revision'].values())
        snapshot = list(objects['snapshot'].values())[0]
        self.maybe_load_snapshot(snapshot)


class RemoteTarLoader(BaseTarLoader):
    """This is able to load from remote/local archive into the swh
       archive.

    This will:

    - create an origin (if it does not exist) and a visit
    - fetch the tarball in a temporary location
    - uncompress it locally in a temporary location
    - process the content of the tarball to persist on swh storage
    - clean up the temporary location

    """
    def prepare(self, *, last_modified, **kwargs):
        """last_modified is the time of last modification of the tarball.

        E.g https://ftp.gnu.org/gnu/8sync/:
            [ ] 8sync-0.1.0.tar.gz	2016-04-22 16:35 	217K
            [ ] 8sync-0.1.0.tar.gz.sig	2016-04-22 16:35 	543
            [ ] ...

        Args:
            origin (dict): Dict with keys {url, type}
            last_modified (str): The date of last modification of the
              archive to ingest.
            visit_date (str): Date representing the date of the
              visit. None by default will make it the current time
              during the loading process.

        """
        self.last_modified = last_modified

    def get_tarball_url_to_retrieve(self):
        return self.origin['url']

    def build_revision(self, filepath, nature, hashes):
        """Build the revision with identifier

        We use the `last_modified` date provided by the caller to
        build the revision.

        """
        return set_original_artifact(
            revision=compute_revision(filepath, self.last_modified),
            filepath=filepath,
            nature=nature,
            hashes=hashes,
        )

    def build_snapshot(self, revision):
        """Build the snapshot targeting the revision.

        """
        branch_name = os.path.basename(self.dir_path)
        return snapshot_from(revision['id'], branch_name)


class LegacyLocalTarLoader(BaseTarLoader):
    """This loads local tarball into the swh archive. It's using the
       revision and branch provided by the caller as scaffolding to
       create the full revision and snapshot (with identifiers).

       This is what's:
       - been used to ingest our 2015 rsync copy of gnu.org
       - still used by the loader deposit

    This will:

    - create an origin (if it does not exist) and a visit
    - uncompress a tarball in a local and temporary location
    - process the content of the tarball to persist on swh storage
    - associate it to a passed revision and snapshot
    - clean up the temporary location

    """
    def prepare(self, *, tar_path, revision, branch_name, **kwargs):
        """Prepare the data prior to ingest it in SWH archive.

        Args:
            tar_path (str): Path to the archive to ingest
            revision (dict): The synthetic revision to associate the
              archive to (no identifiers within)
            branch_name (str): The branch name to use for the
              snapshot.

        """
        self.tar_path = tar_path
        self.revision = revision
        self.branch_name = branch_name

    def get_tarball_url_to_retrieve(self):
        return 'file://%s' % self.tar_path

    def build_revision(self, filepath, nature, hashes):
        """Build the revision with identifier

        We use the revision provided by the caller as a scaffolding
        revision.

        """
        return set_original_artifact(
            revision=self.revision,
            filepath=filepath,
            nature=nature,
            hashes=hashes,
        )

    def build_snapshot(self, revision):
        """Build the snapshot targeting the revision.

        We use the branch_name provided by the caller as a scaffolding
        as well.

        """
        return snapshot_from(revision['id'], self.branch_name)
